spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindImproperSynchronization.java

```diff
@@ -0,0 +1,807 @@
+                        }
+                    } else if (assignedField != null) {
+                        if (isCollection(assignedField)) {
+                            backingCollections.add(xfield, assignedField);
+                            backingCollections.get(assignedField)
+                                    .forEach(f -> backingCollections.add(xfield, f));
+                        }
+                    } else if (isImmutableReturner(item.getSignature())) {
+                        analyzeWrappedField(obj, classContext, location, handle, xfield, 0);
+                    }
+                }
+            }
+
+        }
+    }
+
+    private void analyzeAssignmentsInHierarchy(XField collectionLockObject) throws ClassNotFoundException,
+            CFGBuilderException {
+        String lockDeclaringClassName = collectionLockObject.getClassName();
+        Method[] ownMethods = getThisClass().getMethods();
+
+        for (JavaClass declaringClass : getThisClass().getSuperClasses()) {
+            if (!lockDeclaringClassName.equals(ClassName.toDottedClassName(declaringClass.getClassName()))) {
+                continue;
+            }
+
+            for (Method method : declaringClass.getMethods()) {
+                if (ownMethods.length > 0 && Arrays.asList(ownMethods).contains(method) || method.getCode() == null) {
+                    continue;
+                }
+
+                ClassContext classContext = new ClassContext(declaringClass, AnalysisContext.currentAnalysisContext());
+                analyzeAssignments(method, classContext, collectionLockObject);
+            }
+        }
+    }
+
+    private void checkLockUsageInHierarchy(XField field, XMethod exposingMethod) {
+        try {
+            if (isInherited(field)) {
+                if (declaredInheritedFieldAccessingMethods.get(field).isEmpty()) {
+                    findLockUsageInHierarchy(field).forEach(m -> lockUsingMethodsInHierarchy.add(field, m));
+                }
+                declaredInheritedFieldAccessingMethods.add(field, exposingMethod);
+            }
+        } catch (ClassNotFoundException e) {
+            AnalysisContext.reportMissingClass(e);
+        }
+    }
+
+    private Set<XMethod> findExposureInHierarchy(XField lockObject) throws ClassNotFoundException {
+        Set<XMethod> unsafeMethods = new HashSet<>();
+        Method[] ownMethods = getThisClass().getMethods();
+
+        for (JavaClass declaringClass : getThisClass().getSuperClasses()) {
+            for (Method possibleAccessorMethod : declaringClass.getMethods()) {
+                if (possibleAccessorMethod.getCode() == null) {
+                    continue;
+                }
+
+                if (Arrays.asList(ownMethods).contains(possibleAccessorMethod)
+                        || isInitializerMethod(possibleAccessorMethod.getName())) {
+                    continue;
+                }
```

Comment:
Can you please combine these two `if`s, as the body of these are the same?

Reply:
Yes!